### PR TITLE
qt: add preserve_paths keyword to functions

### PR DESCRIPTION
**Pull request by bruchar1**
supersedes #10995, as it also handle `ui_compile` and `preprocess`, and expose the keyword argument, instead of setting a different default value.

This allow to generate ui and moc under subdirectories, as this is allowed with generic generators.

### CODE REVIEW

File: mesonbuild/modules/qt.py [hunk 1]
```diff
@@ -295,7 +298,7 @@ def has_tools(self, state: 'ModuleState', args: T.Tuple, kwargs: 'HasToolKwArgs'
             required=True,
         ),
         KwargInfo('extra_args', ContainerTypeInfo(list, str), listify=True, default=[]),
-        KwargInfo('method', str, default='auto')
```
Comment (tristan957) [inline mesonbuild/modules/qt.py:298]: Not that it matters, but unnecessary change if you care

**Comment by tristan957** [comment]
> superseeds

Funny misspelling :)

**Comment by eli-schwartz** [comment]
Seems weird to mention in the commit message which other PR this supersedes.

But also I don't see how it is supposed to supersede the other PR?

**Comment by tristan957** [comment]
@bruchar1 it is [supersedes](https://www.merriam-webster.com/dictionary/supersede)

File: test cases/frameworks/4 qt/meson.build [hunk 2]
```diff
@@ -55,11 +55,15 @@ foreach qt : ['qt4', 'qt5', 'qt6']
       method : get_option('method')
     )
     # XML files that need to be compiled with the uic tol.
-    prep += qtmodule.compile_ui(sources : 'mainWindow.ui', method: get_option('method'))
+    prep += qtmodule.compile_ui(
```
Comment (tristan957) [inline test cases/frameworks/4 qt/meson.build:58]: Is it possible to extend the test to also use compile_moc?

Comment (bruchar1) [reply test cases/frameworks/4 qt/meson.build:58]: TBH, I'm not familiar enough with Qt to know now to do a simple test for that. My usecase was for .ui files. I just extended it to moc as well, but in our codebase, it seems the current `moc_compile` is working without modifications.

**Comment by tristan957** [comment]
Seems like this comment is all that is left to answer https://github.com/mesonbuild/meson/pull/11795#issuecomment-1553046787

File: docs/markdown/snippets/qt_preserve_path_from.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+## Added `preserve_path_from` keyword to qt module functions.
+
+In `qt4`, `qt5`, and `qt6` modules, `compile_ui`, `compile_moc`,
+and `preprocess` functions now have a `preserve_path_from` keyword.
+
+If given, it specifies that the output
```
Comment (tristan957) [inline docs/markdown/snippets/qt_preserve_path_from.md:6]: Wrap at 80 column width

**Comment by tristan957** [comment]
@eli-schwartz did you have anything more to say on this PR?

**Comment by tristan957** [comment]
CI failure seems unrelated given it is a fortran test that failed

**Comment by tristan957** [comment]
@xclaesse @eli-schwartz does this PR look good to either of you? Could be a good vacation present to get this one in! :)

**Comment by tristan957** [comment]
I think that was a good change. It matches with `install_headers`

**Comment by punytroll** [comment]
> I just simplified it to use a bool instead of a path. Preserving paths from current source dir is probably the only real usecase. If somebody really needs to specify a different path, we could extend it later (or he can simply use a custom_target)

Thanks for this change. The other solution was, of course, way more general, but, I think, pointlessly so. Thanks, for giving it to the other Qt commands as well!

**Comment by punytroll** [comment]
I've tested this PR for my use case (qt.compile_moc() with headers) as well and it works beautifully. I hope, this makes it into 1.3.

**Comment by bruchar1** [comment]
@eli-schwartz @xclaesse Is there anything preventing merging this?

File: docs/markdown/_include_qt_base.md [hunk 4]
```diff
@@ -49,6 +54,11 @@ It takes no positional arguments, and the following keyword arguments:
   - `dependencies`: dependency objects whose include directories are used by moc.
   - `include_directories` (string | IncludeDirectory)[]: A list of `include_directory()`
     objects used when transpiling the .moc files
+  - `preserve_paths` bool: *New in 1.4.0*. If `true`, specifies that the output
```
Comment (tristan957) [inline docs/markdown/_include_qt_base.md:57]: I feel like we typically use `Since` and not `New in`. Are you matching something else?

Comment (bruchar1) [reply docs/markdown/_include_qt_base.md:57]: This file is full of `New in ...`, but `Since` is good for me!

Comment (tristan957) [reply docs/markdown/_include_qt_base.md:57]: Oh I was trying to say if you're being consistent, that is good with me. But whatever you want.

**Review by dcbaker — APPROVED** [review]
